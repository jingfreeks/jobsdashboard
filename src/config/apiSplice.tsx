/* eslint-disable @typescript-eslint/no-explicit-any */
import {setCredentials, setLogout} from '@/features/auth';
import {createApi, fetchBaseQuery} from '@reduxjs/toolkit/query/react';
const baseUrl = import.meta.env.VITE_API_BASE_URL;
const baseQuery = fetchBaseQuery({
  baseUrl,
  credentials: 'include',
   
  prepareHeaders: (headers, {getState}: any) => {
    const token = getState().auth.token;
    if (token) {
      headers.set('authorization', `Bearer ${token}`);
    }
    return headers;
  },
});

const baseQueryWithAuth = async (
  args: string,
  api: any,
  extraOptions: string,
) => {
  let result: any = await baseQuery(args, api, extraOptions);
  console.log(result);
  if (result?.error?.originalStatus === 403) {
    const refereshResult = await baseQuery('/refresh', api, extraOptions);

    if (refereshResult?.data) {
      const currentAuth = api.getState().auth;
      //store the new token
      api.dispatch(setCredentials({
        user: currentAuth.user,
        accessToken: (refereshResult.data as any).accessToken,
        userId: currentAuth.userId,
        roles: currentAuth.roles
      }));
      //retry the original query with new access token
      result = await baseQuery(args, api, extraOptions);
    } else {
      api.dispatch(setLogout());
    }
  }
  return result;
};

export const apiSlice = createApi<any, any>({
  baseQuery: baseQueryWithAuth,
  reducerPath: 'api',
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  endpoints: _ => ({}),
  tagTypes: ['City', 'Jobs', 'MyJobs', 'JobDetails', 'Profile'] as any,
});