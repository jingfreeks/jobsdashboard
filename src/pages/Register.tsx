import { useNavigate, Link } from 'react-router-dom'
import { useForm } from 'react-hook-form'
import logo from '@/assets/react.svg'
import { useSignupMutation } from '@/features/authSlice';
import { useDispatch } from 'react-redux';
import { setCredentials } from '@/features/auth';
import type { FetchBaseQueryError } from '@reduxjs/toolkit/query';

function isFetchBaseQueryError(error: unknown): error is FetchBaseQueryError {
  return typeof error === 'object' && error != null && 'status' in error;
}

// Add a Facebook SVG icon
const FacebookIcon = () => (
  <svg className="w-5 h-5 mr-2" fill="currentColor" viewBox="0 0 24 24"><path d="M22.675 0h-21.35C.595 0 0 .592 0 1.326v21.348C0 23.408.595 24 1.325 24h11.495v-9.294H9.692v-3.622h3.128V8.413c0-3.1 1.893-4.788 4.659-4.788 1.325 0 2.463.099 2.797.143v3.24l-1.918.001c-1.504 0-1.797.715-1.797 1.763v2.313h3.587l-.467 3.622h-3.12V24h6.116C23.406 24 24 23.408 24 22.674V1.326C24 .592 23.406 0 22.675 0"/></svg>
);

type RegisterFormData = {
  username: string;
  email: string;
  password: string;
};

const Register = () => {
  const navigate = useNavigate()
  const dispatch = useDispatch();
  const { register, handleSubmit, formState: { errors } } = useForm<RegisterFormData>();
  const [signup, { isLoading, error }] = useSignupMutation();

  const onSubmit = async (data: RegisterFormData) => {
    try {
      const result = await signup({email: data.email, password: data.password, username: data.username}).unwrap();
      if (result && typeof result === 'object') {
        dispatch(setCredentials({ ...(result as object), user: data.username }));
      }
      navigate('/dashboard');
    } catch {
      // Error is handled by RTK Query's error state
    }
  }

  const handleFacebookRegister = () => {
    alert('Facebook registration would be implemented here.');
  }

  return (
    <div className="min-h-screen flex items-center justify-center bg-[#0856d1]">
      <div className="w-full max-w-md bg-white rounded-2xl shadow-xl px-8 py-10 flex flex-col items-center">
        <img src={logo} alt="Logo" className="w-20 h-20 mb-6" />
        <h1 className="text-2xl font-bold mb-6 text-center">Create Your Account</h1>
        <form onSubmit={handleSubmit(onSubmit)} className="w-full flex flex-col gap-4">
          <div>
            <label htmlFor="username" className="block text-sm font-medium mb-1">Name</label>
            <input
              id="username"
              type="text"
              {...register('username', { required: 'Name is required' })}
              className="w-full px-4 py-3 border border-gray-200 rounded-lg text-base bg-gray-50 focus:outline-none focus:border-blue-600"
            />
            {errors.username && <span className="text-red-500 text-xs">{errors.username.message as string}</span>}
          </div>
          <div>
            <label htmlFor="email" className="block text-sm font-medium mb-1">Email</label>
            <input
              id="email"
              type="email"
              {...register('email', { required: 'Email is required' })}
              className="w-full px-4 py-3 border border-gray-200 rounded-lg text-base bg-gray-50 focus:outline-none focus:border-blue-600"
            />
            {errors.email && <span className="text-red-500 text-xs">{errors.email.message as string}</span>}
          </div>
          <div>
            <label htmlFor="password" className="block text-sm font-medium mb-1">Password</label>
            <input
              id="password"
              type="password"
              {...register('password', { required: 'Password is required' })}
              className="w-full px-4 py-3 border border-gray-200 rounded-lg text-base bg-gray-50 focus:outline-none focus:border-blue-600"
            />
            {errors.password && <span className="text-red-500 text-xs">{errors.password.message as string}</span>}
          </div>
          {error ? (
            <div className="text-red-500 text-sm mb-2 text-center">
              {isFetchBaseQueryError(error)
                ? (error.data as { error?: { message?: string } })?.error?.message || "Registration failed"
                : "Registration failed"}
            </div>
          ) : null}
          <button type="submit" className="w-full py-3 bg-[#0856d1] text-white rounded-lg font-semibold hover:bg-blue-700 transition mb-2" disabled={isLoading}>Sign up with Email</button>
          <button type="button" onClick={handleFacebookRegister} className="w-full py-3 bg-blue-600 text-white rounded-lg font-semibold flex items-center justify-center hover:bg-blue-700 transition">
            <FacebookIcon />Sign up with Facebook
          </button>
        </form>
        <div className="w-full flex flex-col items-center mt-4 gap-1">
          <span className="text-sm text-gray-700">Already have an account? <Link to="/login" className="text-[#0856d1] hover:underline font-medium">Log in</Link></span>
        </div>
      </div>
    </div>
  )
}

export default Register 