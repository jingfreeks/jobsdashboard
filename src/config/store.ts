import { configureStore } from '@reduxjs/toolkit';
import { persistStore, persistReducer, FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER } from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import authReducer from '@/features/auth';
import {apiSlice} from './apiSplice';

// Configure persistence for auth reducer
const authPersistConfig = {
  key: 'auth',
  storage,
  whitelist: ['user', 'token', 'userId', 'roles'], // Only persist these fields
  timeout: 10000, // 10 second timeout for storage operations
  serialize: true, // Enable serialization
  deserialize: true, // Enable deserialization
};

const persistedAuthReducer = persistReducer(authPersistConfig, authReducer);

export const store = configureStore({
  reducer: {
    auth: persistedAuthReducer,
    [apiSlice.reducerPath]: apiSlice.reducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
        ignoredPaths: ['auth'], // Ignore auth state for serialization checks
      },
      immutableCheck: {
        ignoredPaths: ['auth'], // Ignore auth state for immutability checks
      },
    }).concat(apiSlice.middleware),
  devTools: process.env.NODE_ENV !== 'production',
});

export const persistor = persistStore(store);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch; 