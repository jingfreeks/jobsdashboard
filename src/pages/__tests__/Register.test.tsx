import { describe, it, expect, vi } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import Register from '../Register';
import { store } from '@/config/store';

// Mock useSignupMutation to avoid real API calls
vi.mock('@/features/authSlice', () => ({
  useSignupMutation: () => [vi.fn(), { isLoading: false, error: undefined }],
}));

// Mock setCredentials to avoid Redux side effects
vi.mock('@/features/auth', () => ({
  __esModule: true,
  default: () => ({}), // mock reducer
  setCredentials: vi.fn(),
}));

describe('Register', () => {
  it('renders register form fields and button', () => {
    render(
      <Provider store={store}>
        <MemoryRouter>
          <Register />
        </MemoryRouter>
      </Provider>
    );
    expect(screen.getByLabelText(/Name/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/Email/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/Password/i)).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /Sign up with Email/i })).toBeInTheDocument();
  });

  it('shows validation error if fields are empty', async () => {
    render(
      <Provider store={store}>
        <MemoryRouter>
          <Register />
        </MemoryRouter>
      </Provider>
    );
    fireEvent.click(screen.getByRole('button', { name: /Sign up with Email/i }));
    expect(await screen.findByText(/Name is required/i)).toBeInTheDocument();
    expect(await screen.findByText(/Email is required/i)).toBeInTheDocument();
    expect(await screen.findByText(/Password is required/i)).toBeInTheDocument();
  });
}); 