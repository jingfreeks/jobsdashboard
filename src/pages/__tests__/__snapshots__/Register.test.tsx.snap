// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Register > Basic Rendering > matches register form snapshot 1`] = `
<div>
  <div
    class="min-h-screen flex items-center justify-center bg-[#0856d1]"
  >
    <div
      class="w-full max-w-md bg-white rounded-2xl shadow-xl px-8 py-10 flex flex-col items-center"
    >
      <img
        alt="Logo"
        class="w-20 h-20 mb-6"
        src="mocked-logo.svg"
      />
      <h1
        class="text-2xl font-bold mb-6 text-center"
      >
        Create Your Account
      </h1>
      <form
        class="w-full flex flex-col gap-4"
      >
        <div>
          <label
            class="block text-sm font-medium mb-1"
            for="username"
          >
            Name
          </label>
          <input
            class="w-full px-4 py-3 border border-gray-200 rounded-lg text-base bg-gray-50 focus:outline-none focus:border-blue-600"
            id="username"
            name="username"
            type="text"
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium mb-1"
            for="email"
          >
            Email
          </label>
          <input
            class="w-full px-4 py-3 border border-gray-200 rounded-lg text-base bg-gray-50 focus:outline-none focus:border-blue-600"
            id="email"
            name="email"
            type="email"
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium mb-1"
            for="password"
          >
            Password
          </label>
          <input
            class="w-full px-4 py-3 border border-gray-200 rounded-lg text-base bg-gray-50 focus:outline-none focus:border-blue-600"
            id="password"
            name="password"
            type="password"
          />
        </div>
        <button
          class="w-full py-3 bg-[#0856d1] text-white rounded-lg font-semibold hover:bg-blue-700 transition mb-2"
          type="submit"
        >
          Sign up with Email
        </button>
        <button
          class="w-full py-3 bg-blue-600 text-white rounded-lg font-semibold flex items-center justify-center hover:bg-blue-700 transition"
          type="button"
        >
          <svg
            class="w-5 h-5 mr-2"
            fill="currentColor"
            viewBox="0 0 24 24"
          >
            <path
              d="M22.675 0h-21.35C.595 0 0 .592 0 1.326v21.348C0 23.408.595 24 1.325 24h11.495v-9.294H9.692v-3.622h3.128V8.413c0-3.1 1.893-4.788 4.659-4.788 1.325 0 2.463.099 2.797.143v3.24l-1.918.001c-1.504 0-1.797.715-1.797 1.763v2.313h3.587l-.467 3.622h-3.12V24h6.116C23.406 24 24 23.408 24 22.674V1.326C24 .592 23.406 0 22.675 0"
            />
          </svg>
          Sign up with Facebook
        </button>
      </form>
      <div
        class="w-full flex flex-col items-center mt-4 gap-1"
      >
        <span
          class="text-sm text-gray-700"
        >
          Already have an account? 
          <a
            class="text-[#0856d1] hover:underline font-medium"
            data-discover="true"
            href="/login"
          >
            Log in
          </a>
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`Register > matches register form snapshot 1`] = `
<div>
  <div
    class="min-h-screen flex items-center justify-center bg-[#0856d1]"
  >
    <div
      class="w-full max-w-md bg-white rounded-2xl shadow-xl px-8 py-10 flex flex-col items-center"
    >
      <img
        alt="Logo"
        class="w-20 h-20 mb-6"
        src="/src/assets/react.svg"
      />
      <h1
        class="text-2xl font-bold mb-6 text-center"
      >
        Create Your Account
      </h1>
      <form
        class="w-full flex flex-col gap-4"
      >
        <div>
          <label
            class="block text-sm font-medium mb-1"
            for="username"
          >
            Name
          </label>
          <input
            class="w-full px-4 py-3 border border-gray-200 rounded-lg text-base bg-gray-50 focus:outline-none focus:border-blue-600"
            id="username"
            name="username"
            type="text"
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium mb-1"
            for="email"
          >
            Email
          </label>
          <input
            class="w-full px-4 py-3 border border-gray-200 rounded-lg text-base bg-gray-50 focus:outline-none focus:border-blue-600"
            id="email"
            name="email"
            type="email"
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium mb-1"
            for="password"
          >
            Password
          </label>
          <input
            class="w-full px-4 py-3 border border-gray-200 rounded-lg text-base bg-gray-50 focus:outline-none focus:border-blue-600"
            id="password"
            name="password"
            type="password"
          />
        </div>
        <button
          class="w-full py-3 bg-[#0856d1] text-white rounded-lg font-semibold hover:bg-blue-700 transition mb-2"
          type="submit"
        >
          Sign up with Email
        </button>
        <button
          class="w-full py-3 bg-blue-600 text-white rounded-lg font-semibold flex items-center justify-center hover:bg-blue-700 transition"
          type="button"
        >
          <svg
            class="w-5 h-5 mr-2"
            fill="currentColor"
            viewBox="0 0 24 24"
          >
            <path
              d="M22.675 0h-21.35C.595 0 0 .592 0 1.326v21.348C0 23.408.595 24 1.325 24h11.495v-9.294H9.692v-3.622h3.128V8.413c0-3.1 1.893-4.788 4.659-4.788 1.325 0 2.463.099 2.797.143v3.24l-1.918.001c-1.504 0-1.797.715-1.797 1.763v2.313h3.587l-.467 3.622h-3.12V24h6.116C23.406 24 24 23.408 24 22.674V1.326C24 .592 23.406 0 22.675 0"
            />
          </svg>
          Sign up with Facebook
        </button>
      </form>
      <div
        class="w-full flex flex-col items-center mt-4 gap-1"
      >
        <span
          class="text-sm text-gray-700"
        >
          Already have an account? 
          <a
            class="text-[#0856d1] hover:underline font-medium"
            data-discover="true"
            href="/login"
          >
            Log in
          </a>
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`Register > renders register form fields and button 1`] = `
<div>
  <div
    class="min-h-screen flex items-center justify-center bg-[#0856d1]"
  >
    <div
      class="w-full max-w-md bg-white rounded-2xl shadow-xl px-8 py-10 flex flex-col items-center"
    >
      <img
        alt="Logo"
        class="w-20 h-20 mb-6"
        src="/src/assets/react.svg"
      />
      <h1
        class="text-2xl font-bold mb-6 text-center"
      >
        Create Your Account
      </h1>
      <form
        class="w-full flex flex-col gap-4"
      >
        <div>
          <label
            class="block text-sm font-medium mb-1"
            for="username"
          >
            Name
          </label>
          <input
            class="w-full px-4 py-3 border border-gray-200 rounded-lg text-base bg-gray-50 focus:outline-none focus:border-blue-600"
            id="username"
            name="username"
            type="text"
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium mb-1"
            for="email"
          >
            Email
          </label>
          <input
            class="w-full px-4 py-3 border border-gray-200 rounded-lg text-base bg-gray-50 focus:outline-none focus:border-blue-600"
            id="email"
            name="email"
            type="email"
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium mb-1"
            for="password"
          >
            Password
          </label>
          <input
            class="w-full px-4 py-3 border border-gray-200 rounded-lg text-base bg-gray-50 focus:outline-none focus:border-blue-600"
            id="password"
            name="password"
            type="password"
          />
        </div>
        <button
          class="w-full py-3 bg-[#0856d1] text-white rounded-lg font-semibold hover:bg-blue-700 transition mb-2"
          type="submit"
        >
          Sign up with Email
        </button>
        <button
          class="w-full py-3 bg-blue-600 text-white rounded-lg font-semibold flex items-center justify-center hover:bg-blue-700 transition"
          type="button"
        >
          <svg
            class="w-5 h-5 mr-2"
            fill="currentColor"
            viewBox="0 0 24 24"
          >
            <path
              d="M22.675 0h-21.35C.595 0 0 .592 0 1.326v21.348C0 23.408.595 24 1.325 24h11.495v-9.294H9.692v-3.622h3.128V8.413c0-3.1 1.893-4.788 4.659-4.788 1.325 0 2.463.099 2.797.143v3.24l-1.918.001c-1.504 0-1.797.715-1.797 1.763v2.313h3.587l-.467 3.622h-3.12V24h6.116C23.406 24 24 23.408 24 22.674V1.326C24 .592 23.406 0 22.675 0"
            />
          </svg>
          Sign up with Facebook
        </button>
      </form>
      <div
        class="w-full flex flex-col items-center mt-4 gap-1"
      >
        <span
          class="text-sm text-gray-700"
        >
          Already have an account? 
          <a
            class="text-[#0856d1] hover:underline font-medium"
            data-discover="true"
            href="/login"
          >
            Log in
          </a>
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`Register > shows validation error if fields are empty 1`] = `
<div>
  <div
    class="min-h-screen flex items-center justify-center bg-[#0856d1]"
  >
    <div
      class="w-full max-w-md bg-white rounded-2xl shadow-xl px-8 py-10 flex flex-col items-center"
    >
      <img
        alt="Logo"
        class="w-20 h-20 mb-6"
        src="/src/assets/react.svg"
      />
      <h1
        class="text-2xl font-bold mb-6 text-center"
      >
        Create Your Account
      </h1>
      <form
        class="w-full flex flex-col gap-4"
      >
        <div>
          <label
            class="block text-sm font-medium mb-1"
            for="username"
          >
            Name
          </label>
          <input
            class="w-full px-4 py-3 border border-gray-200 rounded-lg text-base bg-gray-50 focus:outline-none focus:border-blue-600"
            id="username"
            name="username"
            type="text"
          />
          <span
            class="text-red-500 text-xs"
          >
            Name is required
          </span>
        </div>
        <div>
          <label
            class="block text-sm font-medium mb-1"
            for="email"
          >
            Email
          </label>
          <input
            class="w-full px-4 py-3 border border-gray-200 rounded-lg text-base bg-gray-50 focus:outline-none focus:border-blue-600"
            id="email"
            name="email"
            type="email"
          />
          <span
            class="text-red-500 text-xs"
          >
            Email is required
          </span>
        </div>
        <div>
          <label
            class="block text-sm font-medium mb-1"
            for="password"
          >
            Password
          </label>
          <input
            class="w-full px-4 py-3 border border-gray-200 rounded-lg text-base bg-gray-50 focus:outline-none focus:border-blue-600"
            id="password"
            name="password"
            type="password"
          />
          <span
            class="text-red-500 text-xs"
          >
            Password is required
          </span>
        </div>
        <button
          class="w-full py-3 bg-[#0856d1] text-white rounded-lg font-semibold hover:bg-blue-700 transition mb-2"
          type="submit"
        >
          Sign up with Email
        </button>
        <button
          class="w-full py-3 bg-blue-600 text-white rounded-lg font-semibold flex items-center justify-center hover:bg-blue-700 transition"
          type="button"
        >
          <svg
            class="w-5 h-5 mr-2"
            fill="currentColor"
            viewBox="0 0 24 24"
          >
            <path
              d="M22.675 0h-21.35C.595 0 0 .592 0 1.326v21.348C0 23.408.595 24 1.325 24h11.495v-9.294H9.692v-3.622h3.128V8.413c0-3.1 1.893-4.788 4.659-4.788 1.325 0 2.463.099 2.797.143v3.24l-1.918.001c-1.504 0-1.797.715-1.797 1.763v2.313h3.587l-.467 3.622h-3.12V24h6.116C23.406 24 24 23.408 24 22.674V1.326C24 .592 23.406 0 22.675 0"
            />
          </svg>
          Sign up with Facebook
        </button>
      </form>
      <div
        class="w-full flex flex-col items-center mt-4 gap-1"
      >
        <span
          class="text-sm text-gray-700"
        >
          Already have an account? 
          <a
            class="text-[#0856d1] hover:underline font-medium"
            data-discover="true"
            href="/login"
          >
            Log in
          </a>
        </span>
      </div>
    </div>
  </div>
</div>
`;
